src/vehicle.cpp

```diff
@@ -2271,6 +2276,98 @@ void Vehicle::HandleLoading(bool mode)
+	}
+	if (this->cur_implicit_order_index != first_manual_order) return;
+
+	/* A "last departure" >= now means we've already calculated the separation */
+	Ticks now = _date * DAY_TICKS + _date_fract;
+	if (this->first_order_last_departure >= now) return;
+
+	/* Calculate round trip time from last departure and now - automatic separation waiting time is not included */
+	if (this->first_order_last_departure > 0) {
+		this->first_order_round_trip_time = max(0, now - this->first_order_last_departure);
+	}
+
+	/* To work out the automatic separation waiting time we need to know:
+	 *   - When the last vehicle departed or will depart
+	 *   - Average time to perform the order list (as sum/count)
+	 *   - How many vehicles are currently operating the order list
+	 *   - How many vehicles are currently queuing for the first manual order
+	 */
+	Date last_departure = 0;
+	uint round_trip_sum = 0;
+	uint round_trip_count = 0;
+	int vehicles = 0;
+	int vehicles_queuing = 0;
+	Vehicle *v = this->FirstShared();
+	while (v != nullptr) {
+		last_departure = max(last_departure, v->first_order_last_departure);
+		if (v->first_order_round_trip_time > 0) {
+			round_trip_sum += v->first_order_round_trip_time;
+			round_trip_count++;
+		}
+		/* A stopped vehicle is not included; it might be stopped by player or parked in a depot */
+		if (!(v->vehstatus & VS_STOPPED)) {
+			vehicles++;
+			/* Count vehicles queing for the first manual order but not currently in the station */
+			if (v != this && v->cur_speed == 0 && v->cur_implicit_order_index == first_manual_order && !v->current_order.IsType(OT_LOADING)) {
+				vehicles_queuing++;
+			}
+		}
+		v = v->NextShared();
+	}
+
+	/* Calculate the mean round trip time and separation. The time spent queuing for stations is included in vehicle
+	 * round trip times.
+	 *
+	 * For a single shared order into a single station, this will increase and decrease the separation as needed.
+	 * However, for multiple shared orders into the same station, each shared order can back up the others and all
+	 * the routes will slowly increase their separation until every available vehicle is in the same queue.
+	 *
+	 * To counter this, we need to reduce the round trip time when vehicles are queuing. The scaling here reduces it
+	 * by twice the proportion of queuing vehicles, e.g. if 1/N vehicles are queuing, the RTT is reduced by 2/N.
+	 *
+	 * This is based on the idea that if only M/N vehicles are progressing, the non-queuing RTT is approximately M/N
+	 * of the measured RTT, because (N-M)/N of the RTT is spent in the queue, with the 'twice' coming from the need
+	 * to over-compensate rather than aim exactly for the ideal (which is very approximate here). */
+	int round_trip_time = round_trip_count > 0 ? round_trip_sum / round_trip_count : 0;
+	int vehicles_moving_ratio = max(1, vehicles - 2 * vehicles_queuing);
+	int separation = max(1, vehicles > 0 ? ((round_trip_time * vehicles_moving_ratio) / vehicles) / vehicles : 1);
```

Comment:
`a * b / c / d` == `((a * b) / c) / d`. Adding those extra `()` doesn't add anything, but it suggests it does, confusing the reader it is something special. I suggest removing them :)

Reply:
It's been a while since I did C++ like this, thanks for reminded me; parentheses removed.